### DISCUSSION TITLE
Realmの初期データを登録したい

### DISCUSSION THREAD

**Discussion by silli-3072** [lit-iphone-course/lit-iphone-course#33] [Question-original] [answered]
アプリがインストールされた時点でRealm内にデータを入れておきたく[この記事](https://naoya-ono.com/swift/swift-realm-initial-data/)を参考に実装を試みました。**default.realm**という、ファイルがどこにあるかわからずRealm studioを用いてないからかと思い、Realm studioをインストールしファイルを探しましたが見つかりませんでした。
**default.realm**というファイルがどこに存在するのか教えていただきたいです！

**Comment by txxxxc** [comment, marked as answer]
打ったコマンドと結果をコピペで教えてほしい！！

**Reply by silli-3072** [reply]
あ！！ごめんなさい！めちゃくちゃ今更新したのですがRealm studioは開くことができました！！ただ、default.realmを見つけることができませんでした。

**Reply by txxxxc** [reply]
https://sakura-bird1.hatenablog.com/entry/2018/01/24/154425

これ見て試してみてほしいー！

**Reply by silli-3072** [reply]
見つかりました〜〜！

**Comment by silli-3072** [comment]
ファイルを入れた先の部分がうまく実装できなかったので、教えていただきたいです！[この記事](https://naoya-ono.com/swift/swift-realm-initial-data/)のステップ２の部分を記述したら`Invalid redeclaration of ['application(_:didFinishLaunchingWithOptions:)'`とエラーが出てしまい解決できませんでした。記事参考に一階層大きいファイルを入れてみましたがだめでした。どのように実装すれば良いでしょうか？GitHub(https://github.com/silli-3072/LA4_TimeKorean)

**Reply by txxxxc** [reply]
多分application(_:didFinishLaunchingWithOptions:)
って言うメソッドを二回定義しちゃってるので、一つに統合したら動くと思います！！

**Reply by silli-3072** [reply]
ほんとだ！！そこの部分でした！
ただ、そこの部分でnillが出てしまってうまく起動できませんでした...[この記事](https://naoya-ono.com/swift/swift-realm-initial-data/)のファイル名を変えて実装したのですが...
エラー文は`Unexpectedly found nil while unwrapping an Optional value`です

**Reply by txxxxc** [reply]
エラー吐いてるところのコードも見せてほしい！！

**Reply by silli-3072** [reply]
こちらです
`if !FileManager.default.fileExists(atPath: defaultRealmPath.path) {
          do {
            try FileManager.default.copyItem(at: bundleRealmPath!, to: defaultRealmPath)
          } catch let error {
              print("error: \(error)")
            }
        }
`

**Reply by txxxxc** [reply]
bundleRealmPathでアクセスしてるファイルがないって怒られてるっぽいので、ファイル名確認してみてほしい！

**Reply by silli-3072** [reply]
このファイル名じゃないですか...?
<img width="1512" alt="スクリーンショット 2022-09-21 22 31 08" src="https://user-images.githubusercontent.com/105117616/191517259-476a25d6-aed4-4a57-aa8f-4728d82eba64.png">

**Reply by txxxxc** [reply]
合ってる気がする！
一旦手元で試してみたいからpushお願いします！🙌

**Reply by silli-3072** [reply]
お願いします！

**Comment by txxxxc** [comment]
TARGETS > Build Phases > Copy Bundle Resourcesにファイルを追加してあげないとアプリをビルドした時のアウトプットに含まれないのでそこに追加してあげたら普通に動きそう！
一回やってみてほしいー！
![image](https://user-images.githubusercontent.com/22112440/191541675-b61a5629-dee1-401a-8733-bf69e57b8f5e.png)

**Reply by silli-3072** [reply]
やってみたのですが動きません😭同じ部分同じエラー文です

**Reply by txxxxc** [reply]
まじか！
こっちの手元で動くのは確認済みだから、一旦`TARGETS > Build Phases > Copy Bundle Resources`の部分のスクショ送ってもらってい
い？

**Reply by silli-3072** [reply]
こちらです！
<img width="1512" alt="スクリーンショット 2022-09-22 10 02 27" src="https://user-images.githubusercontent.com/105117616/191635900-ad07fd9d-9977-40da-ab3c-4487fedca738.png">

**Reply by txxxxc** [reply]
一回⌘+Shift+Kでキャッシュ削除してビルドしてみてほしいー！

**Reply by silli-3072** [reply]
してもダメでした！ちなみに、アンインストールして再インストールしても無理でした

**Reply by txxxxc** [reply]
あ、コード見てて気づいたんだけど以下の部分多分いらないから消してもいいかも...！(消した段階でpushお願いします！)
https://github.com/silli-3072/LA4_TimeKorean/blob/main/TimeKorean/AppDelegate.swift#L14-L19

あと、default.realmのファイルを一回削除して入れ直すとかしてみてほしいかも...!コードは間違ってないと思うので！

**Reply by silli-3072** [reply]
`return true`がないとエラーが出てしまいます！
入れ直したのですが、ビルドできませんでした。ファイル内にデータは入っていたのですがnillが出ちゃっています。

**Reply by txxxxc** [reply]
あ、return trueは消さなくて大丈夫！

**Reply by txxxxc** [reply]
ここだ！
`let bundleRealmPath = Bundle.main.url(forResource: "default.realm", withExtension: "realm")`

forResourceにはファイル名、withExtensionには拡張子を指定してあげないとなので、これだとdefault.realm.realmを指定してることになってる！
なので、`forResource: "default.realm"`を`forResource: "default"`に変更してあげましょう！

**Reply by silli-3072** [reply]
できました！！
ただ、データを持ち出すところでエラーが出てしまいうまくできませんでした。写真のエラー文なのですが、少し調べてみます。
<img width="1512" alt="スクリーンショット 2022-09-22 11 24 28" src="https://user-images.githubusercontent.com/105117616/191644552-0cf47c0b-3b19-41ab-a158-9cc1f07abc1b.png">

**Reply by txxxxc** [reply]
realmのファイルのバージョンが低いって怒られてるから、ワンチャンRealm Studioのバージョンが低いかもしれないな🤔
ヤバそうだったらまた教えて下さいー！

**Reply by silli-3072** [reply]
昨日入れたRealm Studioなので僕の入れるファイルが悪いかもです...

**Reply by txxxxc** [reply]
https://github.com/realm/realm-studio
こっから最新版いれてやってみるのがいいかもしれん！(てかこっちの手元では普通に動いたからなんでなんやろ笑)

**Reply by silli-3072** [reply]
最新でした...

**Reply by txxxxc** [reply]
うえええまじか...一旦default.realm作り直してみる？

**Reply by silli-3072** [reply]
エラー治すタイミングだったり、初期データ確認するためにアンインストールをしていて今のファイルしか残されていないんですよね...

**Reply by txxxxc** [reply]
既存のdefault.realm消して、新しく作るだけなので今のファイルは関係なさそう(俺なんか勘違いしてるかも...?ごめん)

**Reply by txxxxc** [reply]
あとdefault.realmを使った初期化方法がうまくいかなそうだったら、コードで書くこともできるよ(UserDefaultsで初回起動を判定して初回起動だったらオブジェクト生成して保存する的なやつ)

**Reply by silli-3072** [reply]
初期化の部分のコードってよりかはrealmから読み込む部分でのエラーな気がします。
前、エラー治んなくて、クリーンかけて実行した時も同じエラー出てその時は再インストールしたら治ったので...

**Reply by txxxxc** [reply]
なるほどなーーー
あ、めちゃくちゃ忘れてたんだけど初回起動だけRealmのデータ反映させるロジックって書いてるっけ？
結局その処理は必要そうだし、ワンチャンそれで治るので一旦初回起動だけRealmにデータを突っ込む処理書いてもらっていい？

**Comment by txxxxc** [comment]
あ、そういうことか！
Realmのバージョン低い説あるから、Realm入れ直してみるといいかもしれん！結局バージョンが合ってないのが問題なので！！！
(そういえば毎回手元で試す時にRealm入れ直してたわ)

**Reply by silli-3072** [reply]
ここの部分で最新版にということですか！！？
<img width="1512" alt="スクリーンショット 2022-09-22 14 19 21" src="https://user-images.githubusercontent.com/105117616/191664361-5ff989b5-56ab-4933-8567-14eec00e4bef.png">
してもダメそうです...

**Reply by txxxxc** [reply]
Link Binary With Librariesの部分からパッケージ消して入れ直すやつやってみてほしい！(自分はそれで動いた！)
![image](https://user-images.githubusercontent.com/22112440/191664582-dd5b18e0-ad3f-41cd-b5ed-f8193d585fa3.png)

**Reply by silli-3072** [reply]
え！！てぃーてぃーさんの方はボタン押しても動いているんですか！！？
入れ直してもダメでした...

**Reply by txxxxc** [reply]
昨日の夜の時点で全然動いたんだよな....笑一旦エラー文でググってやってみてもらっていい？
こっちでももう一回試してみるわー

**Reply by txxxxc** [reply]
もう一回試してみてるけど、めっちゃ普通に全画面動くんだよな🤔🤔🤔

個々の画面のスクショ送ってもらっていい？(バージョン確認したい)

![image](https://user-images.githubusercontent.com/22112440/191708043-6015b8b7-b4a0-4a23-ae30-0c455d0a6fae.png)

**Reply by silli-3072** [reply]
ここのスクショでよかったですか！！？
<img width="1512" alt="スクリーンショット 2022-09-22 18 58 48" src="https://user-images.githubusercontent.com/105117616/191717936-68820ee6-ed68-46ce-a785-d14181b9aed9.png">

ちなみにあれから色々やり直して見ましたがダメでした。そして、GitHub上のものをダウンロードしてやってみたらそっちは初期データの方でエラーが出てやれませんでした😭

**Reply by txxxxc** [reply]
一旦初回起動だけデータ突っ込むロジック書くのお願いしたい🙏
必ず必要になると思うので...!

**Reply by txxxxc** [reply]
なんかRealmが２つはいっちゃってるから以下のやつ試してみてほしい！
1. Package DependenciesにあるRealmを全部消す
2. `Add Packages`から `https://github.com/realm/realm-swift`を入力する
3. RealmとRealm Swift入れる
4. Simulatorからアプリを消す

俺はこれで動いた！笑
参考↓
![image](https://user-images.githubusercontent.com/22112440/191732663-cdbb656e-4ec0-4835-861f-5df53039f51a.png)

**Reply by silli-3072** [reply]
えーーー！！！！！できましたできましたできましためっちゃできました！！！！！！ありがとうございますぅぅぅ！！！！！

**Reply by txxxxc** [reply]
🥰🥰🥰
